adal/src/main/java/com/microsoft/aad/adal/AcquireTokenRequest.java

```diff
@@ -0,0 +1,827 @@
+        mBrokerProxy = new BrokerProxy(appContext);
+    }
+
+    /**
+     * Handles the acquire token logic. Will do authority validation first if developer set valiateAuthority to be
+     * true.
+     */
+    void acquireToken(final IWindowComponent activity, final boolean useDialog, final AuthenticationRequest authRequest,
+                      final AuthenticationCallback<AuthenticationResult> authenticationCallback) {
+        final CallbackHandler callbackHandle = new CallbackHandler(getHandler(), authenticationCallback);
+        // Executes all the calls inside the Runnable to return immediately to
+        // user. All UI
+        // related actions will be performed using Handler.
+        Logger.setCorrelationId(authRequest.getCorrelationId());
+        mRequestCorrelationId = authRequest.getCorrelationId();
+        Logger.v(TAG, "Sending async task from thread:" + android.os.Process.myTid());
+        THREAD_EXECUTOR.execute(new Runnable() {
+            @Override
+            public void run() {
+                Logger.v(TAG, "Running task in thread:" + android.os.Process.myTid());
+                try {
+                    // Validate acquire token call first.
+                    acquireTokenValidation(authRequest);
+                    acquireTokenAfterValidation(callbackHandle, activity, useDialog, authRequest);
+                } catch (final AuthenticationException authenticationException) {
+                    callbackHandle.onError(authenticationException);
+                }
+            }
+        });
+    }
+
+    /**
+     * Developer is using refresh token call to do refresh without cache usage.
+     * App context or activity is not needed. Async requests are created, so this
+     * needs to be called at UI thread.
+     */
+    void refreshTokenWithoutCache(final String refreshToken, final AuthenticationRequest authenticationRequest,
+                                  final AuthenticationCallback<AuthenticationResult> externalCallback) {
+        Logger.setCorrelationId(authenticationRequest.getCorrelationId());
+        Logger.v(TAG, "Refresh token without cache");
+
+        final CallbackHandler callbackHandle = new CallbackHandler(getHandler(), externalCallback);
+
+        // Execute all the calls inside Runnable to return immediately. All UI
+        // related actions will be performed using Handler.
+        THREAD_EXECUTOR.execute(new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    // validate acquire token call first.
+                    acquireTokenValidation(authenticationRequest);
+
+                    final AcquireTokenSilentHandler acquireTokenSilentHandler = new AcquireTokenSilentHandler(mContext,
+                            authenticationRequest, mTokenCacheAccessor);
+                    final AuthenticationResult authResult
+                            = acquireTokenSilentHandler.acquireTokenWithRefreshToken(refreshToken);
```

Comment:
we can refactor this using inheritence, and have a base class doing preTokenRequest/performTokenRequest/postTokenRequest